lib/iris/analysis/__init__.py

```diff
@@ -1169,46 +1241,66 @@ def _percentile(data, axis, percent, fast_percentile_method=False, **kwargs):
         masked arrays.
 
     """
-    # Ensure that the target axis is the last dimension.
-    data = np.rollaxis(data, axis, start=data.ndim)
-    shape = data.shape[:-1]
-    # Flatten any leading dimensions.
-    if shape:
-        data = data.reshape([np.prod(shape), data.shape[-1]])
-    # Perform the percentile calculation.
-    if fast_percentile_method:
-        msg = "Cannot use fast np.percentile method with masked array."
-        if ma.is_masked(data):
-            raise TypeError(msg)
-        result = np.percentile(data, percent, axis=-1)
-        result = result.T
+    # Express axis as list of non-negative integers.
+    if not isinstance(axis, Iterable):
+        axis = [axis % data.ndim]
     else:
-        quantiles = np.array(percent) / 100.0
-        result = scipy.stats.mstats.mquantiles(
-            data, quantiles, axis=-1, **kwargs
+        axis = sorted([dim % data.ndim for dim in axis])
+
+    # Get data as a 1D or 2D view with the target axis as the trailing one.
+    end_size = np.prod([data.shape[dim] for dim in axis])
+    untouched_dims = [dim for dim in range(data.ndim) if dim not in axis]
+    shape = [data.shape[dim] for dim in untouched_dims]
+
+    data = data.transpose(untouched_dims + axis)
+
+    if shape or len(axis) > 1:
+        data = data.reshape([np.prod(shape), end_size])
+
+    if not isinstance(percent, Iterable):
+        scalar_percent = True
+        percent = [percent]
+    percent = np.array(percent)
+
+    # Perform the percentile calculation.
+    if lazy:
+        _partial_percentile = functools.partial(
+            _calc_percentile,
+            percent=percent,
+            fast_percentile_method=fast_percentile_method,
+            **kwargs,
+        )
+
+        result = iris._lazy_data.map_complete_blocks(
+            data, _partial_percentile, (-1,), percent.shape
         )
-    if not ma.isMaskedArray(data) and not ma.is_masked(result):
-        result = np.asarray(result)
     else:
-        result = ma.MaskedArray(result)
+        result = _calc_percentile(
+            data,
+            percent,
+            fast_percentile_method=fast_percentile_method,
+            **kwargs,
+        )
 
     # Ensure to unflatten any leading dimensions.
     if shape:
-        if not isinstance(percent, Iterable):
-            percent = [percent]
-        percent = np.array(percent)
         # Account for the additive dimension.
         if percent.shape > (1,):
             shape += percent.shape
         result = result.reshape(shape)
     # Check whether to reduce to a scalar result, as per the behaviour
     # of other aggregators.
-    if result.shape == (1,) and quantiles.ndim == 0:
-        result = result[0]
+    if result.shape == (1,) and scalar_percent:
+        result = np.squeeze(result)
```

Comment:
Changed this to `np.squeeze` because, having forced `percent` to be a list before the call to `_calc_percentile`, this loop was now getting triggered for masked array types with the fast_percentile_method.  If you index a size (1,) masked array, you get a plain numpy array, which was causing failures e.g. [here](https://github.com/SciTools/iris/blob/2dba026c6c37cfd27f5014dbce864965c0e13642/lib/iris/tests/test_analysis.py#L604).

Reply:
I tried replacing `np.squeeze` with `array.item()` and got the following failures:

```
FAILED lib/iris/tests/test_analysis.py::TestAggregators::test_fast_percentile_1d_75_percent_masked_type_no_mask - AssertionError: <class 'numpy.ndarray'> !...
FAILED lib/iris/tests/test_analysis.py::TestAggregators::test_fast_percentile_2d_two_coords_masked_type_no_mask - AssertionError: <class 'numpy.ndarray'> !...
FAILED lib/iris/tests/unit/analysis/test_PERCENTILE.py::Test_aggregate::test_1d_single - AttributeError: 'float' object has no attribute 'shape'
FAILED lib/iris/tests/unit/analysis/test_PERCENTILE.py::Test_aggregate::test_masked_1d_single - AttributeError: 'float' object has no attribute 'shape'
FAILED lib/iris/tests/unit/analysis/test_PERCENTILE.py::Test_lazy_fast_aggregate::test_1d_single - AttributeError: 'Array' object has no attribute 'item'
FAILED lib/iris/tests/unit/analysis/test_PERCENTILE.py::Test_lazy_aggregate::test_1d_single - AttributeError: 'Array' object has no attribute 'item'
FAILED lib/iris/tests/unit/analysis/test_PERCENTILE.py::Test_lazy_aggregate::test_masked_1d_single - AttributeError: 'Array' object has no attribute 'item'
```

Reply:
> It seems to me that if the returned cube is non-scalar, then the "percentile_over_foo" coordinate should be on its dimension. So we should either fix the coordinate for this case (and add a test) or take out the scalar_percent. Do you have a preference?

I just checked what happens if you do similarly with a 2D coordinate

```python
cube = iris.cube.Cube(np.arange(10).reshape(2, 5))
coord1 = iris.coords.DimCoord([2, 3], long_name="foo")
coord2 = iris.coords.DimCoord(range(5), long_name="bar")
cube.add_dim_coord(coord1, 0)
cube.add_dim_coord(coord2, 1)

print(cube.collapsed("foo", iris.analysis.PERCENTILE, percent=[50]))
```

```
unknown / (unknown)                 (bar: 5)
    Dimension coordinates:
        bar                             x
    Scalar coordinates:
        foo                         2, bound=(2, 3)
        percentile_over_foo         50 percent

```

So I think for consistency it makes sense to take out the `scalar_percent` check, as you originally suggested.

Reply:
Yep, I think I agree (though I was swayed by your earlier argument). I was tempted to say that when a percent list of length 1 was provided we should make that a coordinate of length 1, but I'm pretty sure that's what a scalar coordinate is.

So I'm happy to lose `scalar_percent`